### PR TITLE
Batch Reconstruction Of Bellatrix Blocks

### PR DESCRIPTION
**Pull request by nisdas**
**What type of PR is this?**

Feature Improvement

**What does this PR do? Why is it needed?**

For nodes running with blinded blocks enabled, they have to request the payloads from the execution client before
being able to respond to the remote peer who requested the block range.

However currently prysm does sequential json-rpc requests for each block requested from the execution node. This is an extremely inefficient method as we have to wait for geth to marshal and prysm to unmarshal the payload before we can proceed to the next block. With addition to network latency between the consensus and execution client, timeouts can frequently occur due to how we request payloads from the execution client.

This PR instead decides to batch all json-rpc requests into a single call, thereby making it much more efficient for prysm to retrieve all the desired execution payloads from the execution client. 

**Which issues(s) does this PR fix?**

N.A

**Other notes for review**

### CODE REVIEW

File: beacon-chain/execution/engine_client.go [hunk 1]
```diff
@@ -315,6 +319,43 @@ func (s *Service) ExecutionBlockByHash(ctx context.Context, hash common.Hash, wi
 	return result, handleRPCError(err)
 }
 
+// ExecutionBlockByHashes fetches a batch of execution engine blocks by hash by calling
+// eth_blockByHash via JSON-RPC.
+func (s *Service) ExecutionBlockByHashes(ctx context.Context, hashes []common.Hash, withTxs bool) ([]*pb.ExecutionBlock, error) {
```
Comment (terencechain) [inline beacon-chain/execution/engine_client.go:324]:
```suggestion
func (s *Service) ExecutionBlocksByHashes(ctx context.Context, hashes []common.Hash, withTxs bool) ([]*pb.ExecutionBlock, error) {
```

File: beacon-chain/execution/engine_client.go [hunk 2]
```diff
@@ -315,6 +319,43 @@ func (s *Service) ExecutionBlockByHash(ctx context.Context, hash common.Hash, wi
...
+	elems := make([]gethRPC.BatchElem, 0, numOfHashes)
+	execBlks := make([]*pb.ExecutionBlock, 0, numOfHashes)
+	errs := make([]error, 0, numOfHashes)
+	if numOfHashes == 0 {
+		return execBlks, nil
+	}
+	for _, h := range hashes {
+		blk := &pb.ExecutionBlock{}
+		err := error(nil)
+		newH := h
```
Comment (terencechain) [inline beacon-chain/execution/engine_client.go:337]: Is `newH := h` necessary? We worry about mutation?

File: beacon-chain/execution/engine_client.go [hunk 3]
```diff
@@ -315,6 +319,43 @@ func (s *Service) ExecutionBlockByHash(ctx context.Context, hash common.Hash, wi
...
+		execBlks = append(execBlks, blk)
+		errs = append(errs, err)
+	}
+	ioErr := s.rpcClient.BatchCall(elems)
+	if ioErr != nil {
+		return nil, ioErr
+	}
+	for _, e := range errs {
+		if e != nil {
+			return nil, handleRPCError(e)
```
Comment (terencechain) [inline beacon-chain/execution/engine_client.go:353]: If one of the requests fail, do we fail the whole thing?

File: beacon-chain/execution/engine_client.go [hunk 4]
```diff
@@ -361,6 +402,81 @@ func (s *Service) ReconstructFullBellatrixBlock(
...
+		header, err := b.Block().Body().Execution()
+		if err != nil {
+			return nil, err
+		}
+		if header.IsNil() {
+			return nil, errors.New("execution payload header in blinded block was nil")
+		}
+		// Determine if the block is pre-merge or post-merge. Depending on the result,
+		// we will ask the execution engine for the full payload.
+		executionBlockHash := common.BytesToHash(header.BlockHash())
```
Comment (terencechain) [inline beacon-chain/execution/engine_client.go:432]: We can move this to within the `else` right?

File: beacon-chain/execution/engine_client.go (hunk 3 above)
Comment (nisdas) [reply beacon-chain/execution/engine_client.go:353]: Yeap, we do as the whole batch now becomes inconsistent.

File: beacon-chain/execution/engine_client.go (hunk 2 above)
Comment (nisdas) [reply beacon-chain/execution/engine_client.go:337]: If you dont create a new hash value, the hash value referenced will change with each iteration in the loop.

**Review by terencechain — APPROVED** [review]

**Review by rauljordan — APPROVED** [review]

**Pull request merged**